Short names for any_*_space in PSy layer (for #733)

TeranIvy opened the pull request:
- `any_space_N_<variable_name>` is condensed into `aspcN_<variable_name>`
- `any_discontinuous_space_N_<variable_name>` is condensed into `adspcN_<variable_name>`
- Information about the original function space name and variable name is now in the DoFs comment, e.g.
  ```
  ! Initialise number of DoFs for f3 on any_space_1
  ```
- Changes to the  `FunctionSpace` class:
  - Added `short_name` property and the related method to construct it;
  - Added check and error for an unsupported function space;
  - Moved `mangle_fs_name` within the `FunctionSpace` class as it is called via the `FunctionSpace.mangled_name` property.
- Updated tests. This includes adding compilation where missing and appropriate, using `dist_mem` fixture rather than looping `for distmem in [False, True]` and tidying up test examples code (copyrights, kinds, removal of constructors from LFRic kernel examples etc.)

Successfully tested in LFRic on [this branch](https://code.metoffice.gov.uk/trac/lfric/log/LFRic/branches/dev/benshipway/r23330_long_name_branch_for_iva). For example, `basis_any_space_9_chi_longer_on_any_discontinuous_space_1_latitude_w3` (69 characters) becomes `basis_aspc9_chi_longer_on_adspc1_latitude_w3` (44 characters) (the maximum allowed length of a name in Fortran is 63 characters).

TeranIvy commented:
Tests with compilation are all green. Ready for first review from either @rupertford or @arporter.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/testkern_multi_anyw2.f90:
```diff
@@ -1,57 +0,0 @@
-! Copyright (c) 2017-2019, Science and Technology Facilities Council
```
It's nicer if you (remember to) use `git mv` when renaming files.

TeranIvy replied:
I did! I always use `git mv` when renaming files and it is not clear to me why it's so poorly represented in the diffs - very frustrating.

arporter replied:
Annoying!

arporter commented on src/psyclone/tests/test_files/dynamo0p3/testkern_multi_anyw2_mod.f90:
```diff
@@ -0,0 +1,75 @@
+! Author: R. W. Ford, STFC Daresbury Lab
+! Modified: I. Kavcic, Met Office
+
+module testkern_multi_anyw2_mod
```
Please add a comment describing the purpose of this file.

TeranIvy replied:
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/11.2_any_space.f90:
```diff
@@ -1,7 +1,48 @@
+program any_space_example
 
-    call invoke(testkern_any_space_3_type(a))
+  ! Description: single kernel call in an invoke where the a single operator
```
Oops! Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/11.2_any_space.f90:
```diff
@@ -1,7 +1,48 @@
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
```
I removed your name. I can add it back if there is not enough blame going around for you :)

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/11_any_space.f90:
```diff
@@ -1,8 +1,52 @@
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
```
Left just Rupert's name.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/11.2_any_space.f90:
```diff
@@ -1,7 +1,48 @@
-    call invoke(testkern_any_space_3_type(a))
+  ! Description: single kernel call in an invoke where the a single operator
+  ! argument is specified as any_space. 
+  use inf,                      only : operator_type
```
Agreed, we should be correct. Changed.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/11_any_space.f90:
```diff
@@ -1,8 +1,52 @@
-    call invoke(testkern_any_space_1_type(a, rdt, b, c, qr))
+  ! Description: single kernel call in an invoke where the field arguments are
+  ! specified on any_space with basis functions on any_space for quadrature rule.
+  use inf,                      only : field_type, quadrature_xyoz_type
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/19.18_anyspace_stencil_1.f90:
```diff
@@ -8,25 +41,26 @@ program single_kernel_anyspace_stencil
   ! case where we have different fields with the same and different
   ! any_space names in different kernels.
-  use testkern_same_anyspace_stencil_mod, only: testkern_same_anyspace_stencil_type
+  use inf,                                     only: field_type
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/19.19_anyspace_stencil_2.f90:
```diff
@@ -1,25 +1,60 @@
   ! in different kernels. We check that the same field always has the
   ! same stencil dofmap irrespective of the any_space name.
-  use testkern_same_anyspace_stencil_mod, only: testkern_same_anyspace_stencil_type
+  use inf,                                     only: field_type
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/20.0.1_cma_assembly_scalar.f90:
```diff
@@ -35,24 +31,28 @@
   ! argument
-  use columnwise_op_asm_kernel_scalar_mod, only: columnwise_op_asm_kernel_type
-  use inf,      only: field_type
+  use inf,                                 only: operator_type, &
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/20.1.1_cma_apply_same_spaces.f90:
```diff
@@ -2,8 +2,6 @@
 ! BSD 3-Clause License
 !
 ! Copyright (c) 2017, Science and Technology Facilities Council
```
Updated the date and tidied up the code (correct module name, authors, etc.)

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/20.3_cma_assembly_field.f90:
```diff
@@ -2,8 +2,6 @@
 ! BSD 3-Clause License
 !
 ! Copyright (c) 2017, Science and Technology Facilities Council
```
Updated the date and tidied up the code (correct module names, authors, etc.)

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/20.4_cma_assembly_field_same_fs.f90:
```diff
@@ -35,17 +31,20 @@
 program single_invoke_cma
 
   ! Description: single CMA-assembly function specified in an invoke call
+  use inf,      only: operator_type, field_type, columnwise_operator_type
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/21.1_single_invoke_multi_anyw2.f90:
```diff
@@ -1,43 +1,52 @@
   ! any_w2 function spaces in a kernel call
-  use testkern_multi_anyw2, only: testkern_multi_anyw2_type
-  use inf,      only: field_type
+  use inf,                      only: field_type
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/21.4_single_invoke_anyw2_operator.f90:
```diff
@@ -1,45 +1,53 @@
 
   ! Description: test that correct code is produced when we have an
   ! any_w2 function space with an operator in a kernel call
+  use inf,                         only: field_type, operator_type, quadrature_type
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/4.5.1_multikernel_invokes.f90:
```diff
@@ -1,27 +1,57 @@
 program multikernel_invokes_6
 
-  ! Multiple kernel calls within an invoke where the arguments are specified
+  ! Two same kernel calls within an invoke where the arguments are specified
```
Thanks for the suggestion, done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/4.5.1_multikernel_invokes.f90:
```diff
@@ -1,27 +1,57 @@
+  ! Two same kernel calls within an invoke where the arguments are specified
   ! as any_space
 
+  use inf,                      only : field_type, operator_type, quadrature_rule
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/4.5.2_multikernel_invokes.f90:
```diff
@@ -39,16 +39,18 @@ program multikernel_invokes_452
   ! Multiple kernel calls within an invoke where the arguments are specified
   ! as any_space
 
+  use inf,                      only : field_type, operator_type, quadrature_rule
```
Done.

TeranIvy commented on src/psyclone/tests/test_files/dynamo0p3/4.5_multikernel_invokes.f90:
```diff
@@ -1,20 +1,50 @@
   ! Multiple kernel calls within an invoke where the kernels are specified
   ! as any_space
 
+  use inf,                      only : field_type, operator_type, quadrature_rule
```
Done.

TeranIvy commented on src/psyclone/tests/dynamo0p3_basis_test.py:
```diff
@@ -909,7 +913,9 @@ def test_two_eval_diff_space(tmpdir):
     both require evaluators for the same variable declared as being on the
-    same space '''
+    same space.
+
```
Ok, I guess that refers to other modified docstrings in this file so I removed blank lines from them as well.

TeranIvy commented on src/psyclone/tests/dynamo0p3_basis_test.py:
```diff
@@ -921,32 +927,31 @@ def test_two_eval_same_var_same_space(tmpdir):
     assert gen_code.count(
-        "DEALLOCATE (basis_w0_on_any_discontinuous_space_1_f0, "
-        "diff_basis_w1_on_any_discontinuous_space_1_f0)") == 1
+        "DEALLOCATE (basis_w0_on_adspc1_f0, diff_basis_w1_on_adspc1_f0)") == 1
```
There is one already!

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2793,11 +2878,11 @@ def initialise(self, parent):
                 parent.add(CommentGen(parent,
                                       " Initialise number of DoFs for " +
-                                      function_space.mangled_name))
+                                      arg.name + " on " +
```
You're right. I was trying to leave some useful information in the comment as the mangled names of `any_*_space`s have now short FS space names. I will revert it to mangled name, however, I think it would still be useful to have a full function space and argument name when we go "full symbolic" in the PSy-layer so the users can see what field is being looped over. For the future, what do you think about something like
```
" Initialise number of DoFs for <fspace._orig_name> (write arg: <arg.name>)"
```

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5064,7 +5149,7 @@ class DynInvokeSchedule(InvokeSchedule):
     to the base class so it creates the ones we require. '''
 
-    def __init__(self, arg, reserved_names=None):
+    def __init__(self, arg, reserved_names=None, my_invoke=None):
```
Oops, a left-over from some experimentation, sorry! Removed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -427,45 +411,139 @@ def cma_on_space(function_space, arguments):
             # We do not construct the name-mangled name at this point
             # as the full list of kernel arguments may still be under
             # construction.
             self._mangled_name = None
```
You're right. Removed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -427,45 +411,139 @@ def cma_on_space(function_space, arguments):
+        :rtype: str
+
+        '''
+        if self._short_name:
```
No and yes. Removed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -427,45 +411,139 @@ def cma_on_space(function_space, arguments):
+        '''
+        if self._short_name:
+            return self._short_name
+        self._short_name = self._shorten_fs_name()
```
Yes - too much code duplication. Removed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -427,45 +411,139 @@ def cma_on_space(function_space, arguments):
+                        self._orig_name.lower()):
+                    mngl_name = self._short_name + "_" + arg.name
+                    return mngl_name
+        # Raise an error if a function space that has no kernel arguments
```
Done.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -427,45 +411,139 @@ def cma_on_space(function_space, arguments):
+        :rtype: str
+
+        '''
+        fslist = self._orig_name.split("_")
```
I was hoping to avoid extra 'if` clauses but I agree the checks would be useful. Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/lfric_extract_test.py:
```diff
@@ -359,7 +359,9 @@ def test_single_node_dynamo0p3():
 def test_node_list_dynamo0p3():
     ''' Test that applying Extract Transformation on a list of Nodes
-    produces the correct result in Dynamo0.3 API. '''
+    produces the correct result in Dynamo0.3 API.
```
Done.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1379,186 +1397,180 @@ def test_orientation():
     ''' Tests that any_discontinuous_space is implemented correctly in the
     PSy layer when including multiple spaces, operators on same and different
-    "to" and "from" spaces) and basis/differential basis functions '''
+    "to" and "from" spaces) and basis/differential basis functions.
```
Corrected.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1726,7 +1738,9 @@ def test_dyninvoke_arg_for_fs():
     ''' Test that a call to enforce boundary conditions is *not* added
     following a call to the matrix_vector_kernel_type kernel. Boundary
-    conditions are now explicity specified in the Algorithm as required. '''
+    conditions are now explicity specified in the Algorithm as required.
```
Thanks! Corrected.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -2702,11 +2716,28 @@ def test_arg_desc_func_space_tofrom_err():
 
 
+def test_unrecognised_fspace_error():
+    ''' Tests that an error is raised in FunctionSpace initialisation when
```
Yes, another left-over from copying. Removed `mangled_name`.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -2702,11 +2716,28 @@ def test_arg_desc_func_space_tofrom_err():
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    first_invoke = psy.invokes.invoke_list[0]
+    first_kernel = first_invoke.schedule.coded_kernels()[0]
+    with pytest.raises(GenerationError) as excinfo:
```
Changed and updated the error messages.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -2715,37 +2746,59 @@ def test_mangle_no_space_error():
+    fs_name = "any_space_2"
+    short_name = FunctionSpace(fs_name, first_kernel.arguments).short_name
+    mangled_name = FunctionSpace(fs_name, first_kernel.arguments).mangled_name
+    short_name = FunctionSpace(fs_name, first_kernel.arguments).short_name
```
Oops, removed!

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -427,45 +411,139 @@ def cma_on_space(function_space, arguments):
+        '''
+        # List kernel arguments
+        args = self._kernel_args.args
+        # Mangle the function space name for any_*_space
```
Added the check and updated the `test_no_mangle_specified_function_space` to trigger this error with a "real" function space (invalid FS name will trigger the first constructor check).

TeranIvy commented on src/psyclone/tests/dynamo0p3_cma_test.py:
```diff
@@ -828,41 +826,39 @@ def test_cma_apply_indirection_dofmap_error():
-    a kernel that applies a CMA operator '''
+    ''' Test that we generate correct code for a kernel that applies
+    a CMA operator.
+
```
Removed.

TeranIvy commented on src/psyclone/tests/dynamo0p3_lma_test.py:
```diff
@@ -793,12 +799,11 @@ def test_operators():
+        "ndf_adspc1_op_6,op_6_ncell_3d) :: op_6\n"
         "    END SUBROUTINE dummy_code\n"
         "  END MODULE dummy_mod")
     assert output in generated_code
```
This is a kernel stub test so I am not sure how we do that!

TeranIvy commented on src/psyclone/tests/dynamo0p3_lma_test.py:
```diff
@@ -858,10 +865,9 @@ def test_stub_operator_different_spaces():
+    assert "dimension(ndf_w3,ndf_adspc2_op_1,op_1_ncell_3d)" in result
     field_descriptor = metadata.arg_descriptors[0]
     result = str(field_descriptor)
     assert "function_space_to[2]='w3'" in result
```
Again, I am not sure how to check compilation for stubs.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1323,10 +1326,10 @@ def test_loop_fuse_cma(dist_mem):
+        "cbanded_map_aspc2_lma_op1)\n"
         "        !\n"
         "        CALL testkern_code(nlayers, scalar1, "
         "afield_proxy%data, bfield_proxy%data, cfield_proxy%data, "
```
Added and updated the example invoke program to use correct modules.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1426,14 +1429,14 @@ def test_builtin_single_omp_pdo(monkeypatch, annexed, dist_mem):
+            "      DO df=1,undf_aspc1_f2\n"
             "        f2_proxy%data(df) = f1_proxy%data(df)\n"
             "      END DO\n"
             "      !$omp end parallel do") in result
```
Added (plus tidying up the invoke example for correct module names).

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1494,29 +1497,31 @@ def test_builtin_multiple_omp_pdo(monkeypatch, annexed, dist_mem):
+            "      DO df=1,undf_aspc1_f3\n"
             "        f3_proxy%data(df) = ginger\n"
             "      END DO\n"
             "      !$omp end parallel do\n") in result
```
Added (plus tidying up the invoke example for correct module names).

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1556,7 +1561,7 @@ def test_builtin_loop_fuse_pdo(monkeypatch, annexed, dist_mem):
+            "      DO df=1,undf_aspc1_f1\n"
             "        f1_proxy%data(df) = fred\n"
             "        f2_proxy%data(df) = 3.0\n"
             "        f3_proxy%data(df) = ginger\n"
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1611,17 +1616,17 @@ def test_builtin_single_omp_do(monkeypatch, annexed, dist_mem):
             "        f2_proxy%data(df) = f1_proxy%data(df)\n"
             "      END DO\n"
             "      !$omp end do\n"
             "      !$omp end parallel\n") in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1691,28 +1696,28 @@ def test_builtin_multiple_omp_do(monkeypatch, annexed, dist_mem):
             "        f3_proxy%data(df) = ginger\n"
             "      END DO\n"
             "      !$omp end do\n"
             "      !$omp end parallel") in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1772,499 +1777,494 @@ def test_builtin_loop_fuse_do(monkeypatch, annexed, dist_mem):
             "      !$omp end parallel") in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1772,499 +1777,494 @@ def test_builtin_loop_fuse_do(monkeypatch, annexed, dist_mem):
+            "      DO df=1,undf_aspc1_f1\n"
+            "        asum = asum+f1_proxy%data(df)*f2_proxy%data(df)\n"
+            "      END DO\n"
+            "      !$omp end parallel do\n") in code
```
This was a bit tricky as it required adding and editing `scalar_mod.F90` from the LFRic infrastructure. However, we are probably overdue to start checking compilation of scalar reductions!

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1772,499 +1777,494 @@ def test_builtin_loop_fuse_do(monkeypatch, annexed, dist_mem):
+            "      END DO\n"
+            "      !$omp end do\n"
+            "      !$omp end parallel\n") in code
+
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1772,499 +1777,494 @@ def test_builtin_loop_fuse_do(monkeypatch, annexed, dist_mem):
+            "      DO df=1,undf_aspc1_f1\n"
+            "        asum = asum+f1_proxy%data(df)*f2_proxy%data(df)\n"
+            "      END DO\n"
+            "      !$omp end parallel do\n") in code
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1772,499 +1777,494 @@ def test_builtin_loop_fuse_do(monkeypatch, annexed, dist_mem):
+            "      END DO\n"
+            "      !$omp end do\n"
+            "      !$omp end parallel")
+    assert expected_output in result
```
Added and tidied up invoke example (correct module names).

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1772,499 +1777,494 @@ def test_builtin_loop_fuse_do(monkeypatch, annexed, dist_mem):
+            "        asum = asum+l_asum(1,th_idx)\n"
+            "      END DO\n"
+            "      DEALLOCATE (l_asum)\n")
+    assert expected_output in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1772,499 +1777,494 @@ def test_builtin_loop_fuse_do(monkeypatch, annexed, dist_mem):
+            "      END DO\n"
+            "      !$omp end do\n"
+            "      !$omp end parallel")
+    assert expected_output in result
```
Added and tidied up the related invoke example.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1772,499 +1777,494 @@ def test_builtin_loop_fuse_do(monkeypatch, annexed, dist_mem):
+            "        bsum = bsum+l_bsum(1,th_idx)\n"
+            "      END DO\n"
+            "      DEALLOCATE (l_bsum)")
+    assert expected_output in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2326,97 +2326,259 @@ def test_multi_reduction_real_fuse():
+            "        bsum = bsum+f1_proxy%data(df)\n"
+            "      END DO\n"
+            "      !$omp end parallel do\n") in code
+
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2326,97 +2326,259 @@ def test_multi_reduction_real_fuse():
+            "      DO df=1,undf_aspc1_f1\n"
+            "        f1_proxy%data(df) = bsum*f1_proxy%data(df)\n"
+            "      END DO\n"
+            "      !$omp end parallel do\n") in result
```
Added aand tidies up the related invoke example.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2326,97 +2326,259 @@ def test_multi_reduction_real_fuse():
+            "        f1_proxy%data(df) = bsum*f1_proxy%data(df)\n"
+            "      END DO\n"
+            "      !$omp end do\n"
+            "      !$omp end parallel\n") in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2426,87 +2588,80 @@ def test_multi_builtins_red_then_pdo(monkeypatch, annexed):
-                "      !$omp end parallel do\n") in result
+                "      !$omp end parallel do\n")
+        assert code in result
 
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2521,261 +2676,159 @@ def test_multi_builtins_red_then_do(monkeypatch, annexed):
                 "      !$omp end do\n"
-                "      !$omp end parallel\n") in result
+                "      !$omp end parallel\n")
+        assert code in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2521,261 +2676,159 @@ def test_multi_builtins_red_then_do(monkeypatch, annexed):
+            "      DO df=1,undf_aspc1_f1\n"
+            "        asum = asum+f1_proxy%data(df)\n"
+            "      END DO\n"
+            "      !$omp end parallel do\n") in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2521,261 +2676,159 @@ def test_multi_builtins_red_then_do(monkeypatch, annexed):
                 "      END DO\n"
-                "      !$omp end parallel do\n") in result
+                "      !$omp end parallel do\n")
+        assert code in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2785,131 +2838,60 @@ def test_builtins_usual_then_red_fuse_pdo(monkeypatch, annexed):
+                "      END DO\n"
+                "      !$omp end do\n"
+                "      !$omp end parallel\n")
+        assert code in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -2944,51 +2926,285 @@ def test_loop_fuse_error(dist_mem):
+            "      DO th_idx=1,nthreads\n"
+            "        asum = asum+l_asum(1,th_idx)\n"
+            "      END DO\n"
+            "      DEALLOCATE (l_asum)\n") in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -3170,304 +3350,90 @@ def test_reprod_builtins_red_then_usual_do(monkeypatch, annexed):
                 "      DEALLOCATE (l_asum)\n") in result
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -3170,304 +3350,90 @@ def test_reprod_builtins_red_then_usual_do(monkeypatch, annexed):
+                "        " + names["var"] + " = " + names["var"] + "+" +
+                names["lvar"] + "(1,th_idx)\n"
+                "      END DO\n"
+                "      DEALLOCATE (" + names["lvar"] + ")\n") in code
```
Added.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -6242,10 +6210,10 @@ def test_accenterdata_builtin():
+            "      DO df=1,undf_aspc1_f1\n"
             "        f1_proxy%data(df) = 0.0\n"
             "      END DO\n"
             "      !$acc end parallel\n" in output)
```
Added and tidied up the related invoke example.

TeranIvy commented on src/psyclone/tests/dynamo0p3_builtins_test.py:
```diff
@@ -1980,7 +2024,7 @@ def test_sum_X(dist_mem):
+            "      DO df=1,undf_aspc1_f1\n"
             "        asum = asum+f1_proxy%data(df)\n"
             "      END DO")
         assert output in code
```
Added and tidied up the related invoke examples.

TeranIvy commented on src/psyclone/tests/dynamo0p3_builtins_test.py:
```diff
@@ -2492,25 +2544,25 @@ def test_multi_builtin_single_invoke(monkeypatch, annexed, dist_mem):
-            "      DO df=1,undf_any_space_1_f1\n"
+            "      DO df=1,undf_aspc1_f1\n"
             "        f1_proxy%data(df) = asum*f1_proxy%data(df)\n"
             "      END DO\n") in code
```
Added and tidied up the related invoke example.

TeranIvy commented:
Addressed comments and refreshed the branch. Ready for next review from @arporter.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -515,10 +511,22 @@ def _mangle_fs_name(self):
+           VALID_ANY_DISCONTINUOUS_SPACE_NAMES:
+            raise InternalError(
+                "_mangle_fs_name: function space name '{0}' is not one of "
+                "'any_space' or 'any_discontinuous_space' names.".
```
Used and slighly changed the error message to read better.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -541,12 +549,27 @@ def _shorten_fs_name(self):
+            start = "ad"
+        else:
+            raise InternalError(
+                "_shorten_fs_name: function space name '{0}' is not one of "
```
Used and slighly changed the error message to read better.

TeranIvy commented:
Checked and corrected user and developer documentation - this included re-writing `built_ins.rst` as it was very out-of-date. The HTML and PDF documentation builds. Added requested `ANY_*_SPACE` lists in `dynamo0p3.py` and the relevant test in `dynamo0p3_test.py`. Tests with compilation pass. Ready for another review from @arporter.

arporter commented on doc/user_guide/built_ins.rst:
```diff
@@ -118,15 +129,14 @@ Below is an example of a kernel that is consistent with the
-interface. See the :ref:`api-label` section for different ways to
-translate/generate code.
-::
+``solver_mod.x90`` (see Example 3 in :ref:`LFRic examples <examples_lfric>`
```
s/call/called/ above.

TeranIvy replied:
Done.

arporter commented on doc/user_guide/built_ins.rst:
```diff
@@ -135,111 +145,115 @@ the code parser) the differences between the original algorithm code
   which will call the generated PSy layer code.
 
-The existence of a call to a Built-in has made no difference at this point:
+The existence of calls to Built-ins has made no difference at this point:
```
We can use the shortcut of '::' at the end of this line and remove the '::' on the line below.

TeranIvy replied:
Done.

arporter commented on doc/user_guide/built_ins.rst:
```diff
@@ -135,111 +145,115 @@ the code parser) the differences between the original algorithm code
-Built-ins may be implemented in whatever way the generator
-sees fit with no change to the algorithm and kernel layers.
+A vanilla (with no distributed and shared-memory optimisations) version
+of the generated PSy layer is given below. As expected the kernel code is
```
Comma after "As expected" please.

TeranIvy replied:
Done.

arporter commented on doc/user_guide/built_ins.rst:
```diff
@@ -135,111 +145,115 @@ the code parser) the differences between the original algorithm code
-  duplicate arguments and any literals leaving the three fields being
-  passed in;
-* A use statement is added for the each of the new ``CALL invoke_xx``
+  ``CALL invoke_<xx>``. The calls within the ``CALL invoke_jacobi_mass_lump``
```
Suggest: "... specific calls to ``invoke_<xx>`` routines. The kernel calls within the original ``invoke`` are removed, as are ....literals, leaving the five fields and one operator as arguments;"

TeranIvy replied:
Thanks for the suggestion. Done.

arporter commented on doc/user_guide/psyke.rst:
```diff
@@ -241,26 +241,26 @@ to insert the extract region. As shown below, all children of an
               upper_bound='ndofs']
-	  ...
-	  Schedule[]
+      ...
```
This line is not indented correctly - it should line up with the `1: ` above.

TeranIvy replied:
I realised I messed up indenting when replacing tabs with spaces in this file. Hopefully corrected now.

TeranIvy commented:
Corrected and amended documentation as requested by the reviewer. `psyke.rst` should now show identical indentation when compared to master in browser. Travis permitting, this is ready for the next review.
